core/templates/dev/head/services/SpeechSynthesisChunkerService.js

```diff
@@ -0,0 +1,156 @@
+    // back autogenerated audio.
+    var PUNCTUATION_MARKS_TO_END_CHUNKS = '.!?:;';
+    // Punctuation marks that we don't want to influence the playing back of
+    // autogenerated audio.
+    var PUNCTUATION_MARKS_TO_IGNORE = '"';
+
+    var _speechUtteranceChunker = function(utt, settings, callback) {
+      settings = settings || {};
+      var newUtt;
+      var txt = (settings && settings.offset !== undefined ?
+          utt.text.substring(settings.offset) : utt.text);
+      if (utt.voice && utt.voice.voiceURI === 'native') {
+        // Not part of the spec
+        newUtt = utt;
+        newUtt.text = txt;
+        newUtt.addEventListener('end', function () {
+          if (_speechUtteranceChunker.cancel) {
+            _speechUtteranceChunker.cancel = false;
+          }
+          if (callback !== undefined) {
+            callback();
+          }
+        });
+      }
+      else {
+        var chunkLength = (settings && settings.chunkLength) || 160;
+        var pattRegex = new RegExp('^[\\s\\S]{' +
+          Math.floor(chunkLength / 4) + ',' + chunkLength + '}[' +
+          PUNCTUATION_MARKS_TO_END_CHUNKS + ']{1}|^[\\s\\S]{1,' +
+          chunkLength + '}$|^[\\s\\S]{1,' + chunkLength + '} ');
+        var chunkArr = txt.match(pattRegex);
+        if (chunkArr == null ||
+            chunkArr[0] === undefined ||
+            chunkArr[0].length <= 2) {
+          // Call once all text has been spoken.
+          if (callback !== undefined) {
+            callback();
+          }
+          return;
+        }
+        var chunk = chunkArr[0];
+        newUtt = new SpeechSynthesisUtterance(chunk);
+        var x;
+        for (x in utt) {
+          if (x !== 'text') {
+            newUtt[x] = utt[x];
+          }
+        }
+        newUtt.onend = function() {
+          if (_speechUtteranceChunker.cancel) {
+            _speechUtteranceChunker.cancel = false;
+            return;
+          }
+          settings.offset = settings.offset || 0;
+          settings.offset += chunk.length;
+          _speechUtteranceChunker(utt, settings, callback);
+        }
+      }
+   
+      if (settings.modifier) {
+        settings.modifier(newUtt);
+      }
+
+      // IMPORTANT!! Do not remove: Logging the object out fixes some onend
+      // firing issues. Placing the speak invocation inside a callback
+      // fixes ordering and onend issues.
+      console.log(newUtt);
+      $timeout(function () {
+        speechSynthesis.speak(newUtt);
+      }, 0);
+    };
+
+    var _preprocessHtmlForChunking = function(html) {
+      // Replace certain HTML elements with periods to indicate
+      // pauses in speaking. Also, for some reason, there's a lot
+      // of whitespace (like hundreds of characters) so we trim
+      // it off to avoid blank chunks.
+      html = html.replace(new RegExp('</li>', 'g'), '.').trim();
+      // Strip away HTML tags.
+      var tmp = document.createElement('div');
+      tmp.innerHTML = html;
```

Comment:
In general we eschew using innerHTML; there are security concerns. Can you find another way to do this? There's an rteHelperService in app.js that uses .html(), perhaps that might be better. Also maybe it's worth adding a convertToReadableText() thing in there (see previous comment) --- though I'm not 100% sure if that's the best place for it or if having a new service is better.

Reply:
used html()